### ISSUE TITLE
System Managed Identity to access Graph

### ISSUE THREAD

**Issue by mabster** [AzureAD/microsoft-identity-web#1664]
Hi gang,

Hope this is the right place to ask a question. I'm a little lost here.

I have deployed an ASP.NET Core Web API to an Azure App Service, and given that App Service its own system-assigned managed identity.

I have granted User.Read.All and Directory.Read.All Graph application permissions to the system identity service principal.

What do I need to do to make Microsoft.Identity.Web.MicrosoftGraph use my system managed identity when calling Graph? I have this in my Program.cs:

```
builder.Services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
    .AddMicrosoftIdentityWebApi(builder.Configuration.GetSection("AzureAd"))
    .EnableTokenAcquisitionToCallDownstreamApi()
    .AddMicrosoftGraphAppOnly(auth => new Microsoft.Graph.GraphServiceClient(auth))
    .AddInMemoryTokenCaches();
```

... and my appsettings.json has an AzureAd section like this:

```
  "AzureAd": {
    "Instance": "https://login.microsoftonline.com/",
    "Domain": "<my org>.onmicrosoft.com",
    "TenantId": "<my org's tenant id>",
    "ClientId": "<my app registration client id>",
    "Scopes": "access_as_user",
    "CallbackPath": "/signin-oidc"
  },
```

Should it "just work", or is there some extra step I'm missing? Or are system-assigned MSIs not supported? I've seen a reference in the code to a `UserAssignedManagedIdentityClientId` option, but I've tried setting that to a few different values with no success.

Cheers,
Matt

**Comment by mabster** [comment]
Ok, I got a stack trace.

My request path (/api/users/{id}) is the controller endpoint that is getting the photo for the requested user. So it needs User.Read.All.

I've replaced my tenant id below with {tenant} etc, and for some reason the error contains the _ObjectId_ (not the ClientID) of my system-assigned managed identity, so I have replaced that with {msi-objectid}.

I hope this helps! I'm sure it's something I'm doing wrong but I have no idea what to try next.


```
RequestPath: /api/users/fa7d5f4c-41ad-4c94-8100-ee9dc3a25e7c

An unhandled exception has occurred while executing the request.

Exception: 
Status Code: 0
Microsoft.Graph.ServiceException: Code: generalException
Message: An error occurred sending the request.

 ---&gt; MSAL.NetCore.4.42.0.0.MsalServiceException: 
	ErrorCode: invalid_request
Microsoft.Identity.Client.MsalServiceException: AADSTS70021: No matching federated identity record found for presented assertion. Assertion Issuer: 'https://login.microsoftonline.com/{tenant}/v2.0'. Assertion Subject: '{msi-objectid}'. Assertion Audience: 'fb60f99c-7a34-4190-8149-302f77469936'. https://docs.microsoft.com/en-us/azure/active-directory/develop/workload-identity-federation
Trace ID: 7862bb35-0bdd-40c9-9c7b-cb12a8a0f200
Correlation ID: 97f749c8-f91d-434e-9ecd-111c65037399
Timestamp: 2022-03-23 05:11:08Z
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.HandleTokenRefreshErrorAsync(MsalServiceException e, MsalAccessTokenCacheItem cachedAccessTokenItem)
   at Microsoft.Identity.Client.Internal.Requests.ClientCredentialRequest.ExecuteAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.Internal.Requests.RequestBase.RunAsync(CancellationToken cancellationToken)
   at Microsoft.Identity.Client.ApiConfig.Executors.ConfidentialClientExecutor.ExecuteAsync(AcquireTokenCommonParameters commonParameters, AcquireTokenForClientParameters clientParameters, CancellationToken cancellationToken)
   at Microsoft.Identity.Web.TokenAcquisitionAuthenticationProvider.AuthenticateRequestAsync(HttpRequestMessage request)
   at Microsoft.Graph.AuthenticationHandler.SendAsync(HttpRequestMessage httpRequestMessage, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.&lt;SendAsync&gt;g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
   at Microsoft.Graph.HttpProvider.SendRequestAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
	StatusCode: 400 
	ResponseBody: {"error":"invalid_request","error_description":"AADSTS70021: No matching federated identity record found for presented assertion. Assertion Issuer: 'https://login.microsoftonline.com/{tenant}/v2.0'. Assertion Subject: '{msi-objectid}'. Assertion Audience: 'fb60f99c-7a34-4190-8149-302f77469936'. https://docs.microsoft.com/en-us/azure/active-directory/develop/workload-identity-federation\r\nTrace ID: 7862bb35-0bdd-40c9-9c7b-cb12a8a0f200\r\nCorrelation ID: 97f749c8-f91d-434e-9ecd-111c65037399\r\nTimestamp: 2022-03-23 05:11:08Z","error_codes":[70021],"timestamp":"2022-03-23 05:11:08Z","trace_id":"7862bb35-0bdd-40c9-9c7b-cb12a8a0f200","correlation_id":"97f749c8-f91d-434e-9ecd-111c65037399","error_uri":"https://login.microsoftonline.com/error?code=70021"} 
	Headers: Cache-Control: no-store, no-cache
```

**Comment by mabster** [comment]
The line of code it's crashing on in my controller is:

```
var photo = await _graph.Users[id].Photo.Content.Request().GetAsync();
```

The line works just fine if I supply a ClientSecret, but I didn't think I'd have to with managed identities.

**Comment by mabster** [comment]
I could only get this far by setting the ``AzureAd__UserAssignedManagedIdentityClientId`` environment variable to the app ID of my App Service / System Managed Identity, by the way. If I didn't set that I get a different error - "One client credential type required either: ClientSecret, Certificate OR ClientAssertion must be defined when creating a Confidential Client."

So I guess it will only try to use a managed identity if it's supplied a client ID for the MSI.

**Comment by mabster** [comment]
https://stackoverflow.com/questions/71594455/call-graph-from-azure-app-service-using-managed-identity

I've also tried posting this on Stack Overflow in case someone else is having any luck.

**Comment by mabster** [comment]
This doc:

https://docs.microsoft.com/en-us/azure/app-service/scenario-secure-app-access-microsoft-graph-as-app?tabs=azure-powershell#call-microsoft-graph

... shows how to use `ManagedIdentityCredential` with `GraphServiceClient`, which is what I want, but how do I introduce that code into the simple extension-method based setup code I currently have:

```
builder.Services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
    .AddMicrosoftIdentityWebApi(builder.Configuration.GetSection("AzureAd"))
    .EnableTokenAcquisitionToCallDownstreamApi()
    .AddMicrosoftGraphAppOnly(auth => new Microsoft.Graph.GraphServiceClient(auth))
    .AddInMemoryTokenCaches();
```

... in such a way that the token is cached and refreshed properly? I'm a bit lost here guys. :(

**Comment by mabster** [comment]
OK, GitHub helps those who help themselves.

I've currently got this code in my Program.cs, and it's working at least when running locally in Visual Studio:

```
builder.Services.AddSingleton<ChainedTokenCredential>(services =>
{
    var options = new MicrosoftIdentityOptions();
    builder.Configuration.Bind("AzureAd", options);
    var secretCred = new ClientSecretCredential(options.TenantId, options.ClientId, options.ClientSecret);
    var msiCred = new ManagedIdentityCredential();
    return new ChainedTokenCredential(msiCred, secretCred);
});

// Add services to the container.
builder.Services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
    .AddMicrosoftIdentityWebApi(builder.Configuration.GetSection("AzureAd"))
    .EnableTokenAcquisitionToCallDownstreamApi()
    //.AddMicrosoftGraphAppOnly(auth => new Microsoft.Graph.GraphServiceClient(auth))
    .AddInMemoryTokenCaches();

builder.Services.AddScoped(services =>
{
    var cache = services.GetRequiredService<IMemoryCache>();
    if (!cache.TryGetValue<AccessToken>("graph", out var token))
    {
        var credential = services.GetRequiredService<ChainedTokenCredential>();
        token = credential.GetToken(new TokenRequestContext(new[] { "https://graph.microsoft.com/.default" }));

        cache.Set("graph", token, token.ExpiresOn);
    }
    var accessToken = token.Token;

    return new GraphServiceClient(
        new DelegateAuthenticationProvider(requestMessage =>
        {
            requestMessage.Headers.Authorization = new AuthenticationHeaderValue("bearer", accessToken);
            return Task.CompletedTask;
        }));
});
```

Can someone from this project (all of whom are infinitely more experienced with identity, tokens etc than I) critique this code? Am I on the right track?

**Comment by mabster** [comment]
The code above needed some work, but I have it working now! I've posted my most up-to-date code as an answer on my Stack Overflow question:

https://stackoverflow.com/questions/71594455/call-graph-from-azure-app-service-using-managed-identity/71609780#71609780

Still would love to have someone critique it and suggest a better way.

**Comment by jmprieur** [comment]
@mabster 
- do you use App Services authentication? Or just app services?
- Did you try UserAssignedManagedIdentityClientId  in the AzureAd section of the appsettings.json?

**Comment by mabster** [comment]
Hi @jmprieur!

Just App Services - no authentication at that level. The app itself handles the auth.

I didn't have any luck with UserAssignedManagedIdentityClientId. I tried giving it the value of my system MSI and also tried creating a user-assigned MSI and giving it that ID's object & client ID with no joy.

From Visual Studio I am using User Secrets to manage the ClientSecret so it is merged into the appsettings.json at runtime.

My final (working) code is in that Stack Overflow answer that I've linked to. Hopefully I'm doing everything right at this stage.
